File: apps/api/app/db/models/billing.py
import uuid
import secrets
from datetime import datetime, date

from sqlalchemy import Column, Date, DateTime, Integer, String, Text
from sqlalchemy.dialects.postgresql import JSONB, UUID

from app.db.base import Base, TenantMixin


class SchoolInvoice(Base, TenantMixin):
    __tablename__ = "school_invoices"

    invoice_number = Column(String(50), nullable=False)          # FACT-202509-0626
    created_by = Column(UUID(as_uuid=True), nullable=False)      # admin or tutor user_id

    # Snapshot of student/parent at invoice time (frozen for PDF)
    student_id = Column(UUID(as_uuid=True), nullable=False)
    student_name = Column(String(200), nullable=False)
    student_class = Column(String(100), nullable=True)           # "PS/MS/GS P, Externe libre"
    parent_name = Column(String(200), nullable=True)
    parent_address = Column(JSONB, default=dict)                 # {line1, line2, city, postal_code}

    academic_year = Column(String(20), nullable=False)           # "2025-2026"
    issue_date = Column(Date, nullable=False, default=date.today)
    status = Column(String(20), default="draft")                 # draft|sent|paid|cancelled

    # Line items: [{description, qty, unit_price_cents, total_cents}]
    # qty and unit_price_cents can be negative for discounts
    line_items = Column(JSONB, default=list)

    subtotal_cents = Column(Integer, nullable=False, default=0)
    previous_balance_cents = Column(Integer, nullable=False, default=0)   # negative = credit
    total_due_cents = Column(Integer, nullable=False, default=0)

    # Payment schedule: [{date: "2025-10-15", amount_cents: 5900}]
    payment_schedule = Column(JSONB, default=list)
    payment_method = Column(String(100), nullable=True)     # "Prélèvement", "Virement", etc.
    payment_reference = Column(String(100), nullable=True)  # "4111LACOUTURE"
    bank_account = Column(String(50), nullable=True)        # IBAN shown on invoice

    contact_info = Column(Text, nullable=True)              # free text footer note
    notes = Column(Text, nullable=True)
    paid_cents = Column(Integer, nullable=False, default=0)
    enrollment_request_id = Column(UUID(as_uuid=True), nullable=True)

File: apps/web/src/api/billing.ts
import client from './client';

export interface LineItem {
  description: string;
  qty: number;
  unit_price_cents: number;
  total_cents: number;
}

export interface PaymentScheduleEntry {
  date: string;
  amount_cents: number;
}

export interface InvoiceData {
  id: string;
  tenant_id: string;
  invoice_number: string;
  created_by: string;
  student_id: string;
  student_name: string;
  student_class: string | null;
  parent_name: string | null;
  parent_address: Record<string, string> | null;
  academic_year: string;
  issue_date: string;
  status: string;
  line_items: LineItem[];
  subtotal_cents: number;
  previous_balance_cents: number;
  total_due_cents: number;
  payment_schedule: PaymentScheduleEntry[];
  payment_method: string | null;
  payment_reference: string | null;
  bank_account: string | null;
  contact_info: string | null;
  notes: string | null;
  paid_cents: number;
  enrollment_request_id: string | null;
}

export interface InvoiceCreate {
  student_id: string;
  student_name: string;
  student_class?: string;
  parent_name?: string;
  parent_address?: { line1?: string; line2?: string; postal_code?: string; city?: string };
  academic_year: string;
  issue_date?: string;
  line_items: LineItem[];
  previous_balance_cents?: number;
  payment_schedule?: PaymentScheduleEntry[];
  payment_method?: string;
  payment_reference?: string;
  bank_account?: string;
  contact_info?: string;
  notes?: string;
}

export const getInvoices = (params?: { student_id?: string; status?: string }) =>
  client.get<InvoiceData[]>('/v1/billing/invoices', { params });

export const createInvoice = (data: InvoiceCreate) =>
  client.post<InvoiceData>('/v1/billing/invoices', data);

export const updateInvoice = (id: string, data: { status?: string; notes?: string }) =>
  client.patch<InvoiceData>(`/v1/billing/invoices/${id}`, data);

export const fetchInvoicePdf = (id: string) =>
  client.get<Blob>(`/v1/billing/invoices/${id}/pdf`, { responseType: 'blob' });

export const payFromWallet = (id: string, amount_cents: number) =>
  client.post<InvoiceData>(`/v1/billing/invoices/${id}/pay-from-wallet`, { amount_cents });
